kadai1/kujiraf/converter/converter.go

```diff
@@ -0,0 +1,211 @@
+	for _, ext := range tmp {
+		if !strings.Contains(*ext, ".") {
+			*ext = "." + *ext
+		}
+		if ok := isSupported(*ext); !ok {
+			return fmt.Errorf("%s is not supported", *ext)
+		}
+	}
+
+	// -fromと-toが同じ値ではないか
+	if c.From == c.To {
+		return fmt.Errorf("-from and -to are same. -from %s, -to %s", c.From, c.To)
+	}
+	if (c.From == JPEG || c.From == JPG) && (c.To == JPEG || c.To == JPG) {
+		return fmt.Errorf("-from and -to are same. -from %s, -to %s", c.From, c.To)
+	}
+
+	return nil
+}
+
+func isSupported(ext string) bool {
+	switch ext {
+	case JPEG, JPG, PNG, GIF:
+		return true
+	default:
+		return false
+	}
+}
+
+// DoConvert converts image's extension from c.From to c.To.
+func (c *Converter) DoConvert() error {
+
+	// 出力先パスを絶対パスに変える
+	if !filepath.IsAbs(c.Dst) {
+		abs, err := filepath.Abs(c.Dst)
+		if err != nil {
+			return err
+		}
+		c.Dst = abs
+	}
+	c.debugf("output root path : %s\n", c.Dst)
+
+	// 処理対象のディレクトリに移動する。処理が終われば元の場所に戻る。
+	prevDir, err := filepath.Abs(".")
+	c.debugf("current dir : %s\n", prevDir)
+	if err != nil {
+		return err
+	}
+	os.Chdir(c.Src)
+	defer os.Chdir(prevDir)
+
+	// 処理対象のディレクトリ名取得
+	workDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	srcRoot := filepath.Base(workDir)
+	c.debugf("src dir name : %s\n", srcRoot)
+
+	// 再帰的に処理を実行する
+	err = filepath.Walk(".",
+		func(path string, info os.FileInfo, err error) error {
+			if filepath.Ext(path) == c.From {
+				c.debugf("found. %s\n", path)
+
+				// 対象のファイルに対して処理を実行する
+				err := c.convert(path, srcRoot)
+				if err != nil {
+					return err
+				}
+
+			}
+			return nil
+		})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Converter) convert(inputfile string, root string) error {
+	c.debugf("target file path=%v\n", inputfile)
+
+	// 出力先ディレクトリの作成
+	dst := filepath.Dir(filepath.Join(c.Dst, root, inputfile))
+	c.debugf("output path=%v\n", dst)
+	err := os.MkdirAll(dst, os.ModeDir)
+	if err != nil {
+		return err
+	}
+
+	// ファイルのデコード
+	img, err := c.decode(inputfile)
+	if err != nil {
+		return err
+	}
+
+	// 出力ファイル名取得
+	extlen := len(inputfile) - len(filepath.Ext(inputfile))
+	filename := filepath.Base(inputfile[:extlen]) + c.To
+	outputfile := filepath.Join(dst, filename)
+
+	// ファイルを出力先にコピー
+	err = c.encode(outputfile, img)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("[INFO]conversion complete. converted file from %s to %s\n", inputfile, outputfile)
+	return nil
+}
+
+func (c Converter) decode(input string) (image.Image, error) {
+	in, err := os.Open(input)
+	defer in.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	switch c.From {
+	case PNG:
+		c.debugf("decode %s file %s\n", PNG, input)
+		return png.Decode(in)
+	case GIF:
+		c.debugf("decode %s file %s\n", GIF, input)
+		return gif.Decode(in)
+	default:
+		c.debugf("decode %s file %s\n", JPEG, input)
+		return jpeg.Decode(in)
+	}
+}
+
+func (c Converter) encode(output string, m image.Image) error {
+	newfile, err := os.Create(output)
+	defer newfile.Close()
```

Comment:
こちらもdefer Closeはエラー処理の後が良いかと思います。

Reply:
書き込みの後のCloseもエラーを返すみたいです。

Reply:
修正しました。
Closeのエラーハンドリングも追加しました。

Reply:
Closeのエラーハンドリングはシンプルに以下のようにすれば良いかなと思いましたが、いかがでしょうか。
```go
func do() (err error) {
  f, err := os.Open("book.txt")
  if err != nil {
    return err
  }
  defer func() {
    if ferr := f.Close(); ferr != nil {
      err = ferr
    }
  }()
  // ..code...
  return nil
}
```
cf. https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1

Reply:
コメントありがとうございます。
実は最初、上記のように修正しようと思ったのですが、defer後の`// ..code...`で省略されている処理にてエラーが発生した場合、defer内のClose処理で、前のエラーが上書きされてしまうので、今のような修正に至りました。
このあたりちょっと調べてみましたが、ご紹介いただいた参考リンクの中に、
```
You can also use this package to wrap multiple errors.
This may be necessary, because, f.Close inside defer may swallow any errors before it.
```
とありました。this packageとは[github.com/pkg/errors](https://godoc.org/github.com/pkg/errors)です。
こちら、前のエラーを新しいエラーでラップしてくれるパッケージのようです。
また、[gopher dojoの講義](https://youtu.be/cl8OJCQhRu8?list=PL9MOSAifWs3whvWOsObk3uCBXtVAhD2A7&t=1199)では、このパッケージをもとにerrors.Unwrapが標準パッケージに導入されたと説明があり、fmt.Errorfでエラーをラップしてあげると良いようでした。

Reply:
> defer内のClose処理で、前のエラーが上書きされてしまうので

確かにそのとおりですね。失礼しました。

何が気になっていたかというと、
https://github.com/kujiraf/gopherdojo-studyroom/pull/2/files#diff-b15c93f13a285aa26dcc876eea9a03c89c6f3cacab90069d853d9a771028581cR217
main関数以外で標準出力にエラーを出力していたところなので、ここで前のエラーをラップできれば良さそうですね。

Reply:
ラップできることを知らなかったことと、最大エラー数が２だったのでこれで妥協していました。
一旦セルフマージ後、エラーをラップします。
ありがとうございました！